define record() more explicitly that it is a "set the value"

ktoso opened the pull request:
Depends on https://github.com/apple/swift-metrics/pull/31 so please merge that first

Resolves https://github.com/apple/swift-metrics/issues/34

Only commit to look at here is https://github.com/apple/swift-metrics/commit/22dd2161da89e9beca2d99245a4da9cb4f1d1703

The text is a bit dry, but since don't have a "spec" and the types are the spec basically I figured explaining it a bit more could be fine.

Yasumoto commented on Sources/CoreMetrics/Metrics.swift:
```diff
@@ -160,6 +136,10 @@ public class Recorder {
 
     /// Record a value.
     ///
+    /// Recording a value is meant to have "set" semantics, rather than "add" semantics.
```
```suggestion
    /// Recording a value is meant to have "set" semantics, rather than "add" semantics.
    /// This means that the value of this `Recorder` will match whatever you pass in here.
```

Took a swing at clarifying `"set" semantics` which I understand, but not sure that's immediately clear without some extra context to all folks.

MrLotU replied:
I personally think just defining `"set"` semantics is clear enough.

tomerd replied:
this looks good to me

tomerd commented:
@ktoso can you rebase so easier to review the actual change

tomerd commented on Sources/CoreMetrics/Metrics.swift:
```diff
@@ -136,6 +136,10 @@ public class Recorder {
     /// Record a value.
     ///
+    /// Recording a value is meant to have "set" semantics, rather than "add" semantics.
+    /// Metrics implementations may choose to provide additional "add" or other operations meant to modify an existing value,
```
this seems out of place in the user facing API, maybe move to the Hander API which is aimed for the backend dev?

tomerd commented on Sources/CoreMetrics/Metrics.swift:
```diff
@@ -145,6 +149,10 @@ public class Recorder {
     /// Record a value.
     ///
+    /// Recording a value is meant to have "set" semantics, rather than "add" semantics.
+    /// Metrics implementations may choose to provide additional "add" or other operations meant to modify an existing value,
```
this seems out of place in the user facing API, maybe move to the Hander API which is aimed for the backend dev?

MrLotU replied:
I think it somewhat makes sense to have this here, for example with Prometheus, if you needed/wanted `add` semantics, you would know that 1) `.record()` has `set` semantics and that 2) you could look into if the underlying `PrometheusClient` does have an `add` operation (which in the case of Prometheus, it does)

tomerd replied:
while i fully recognize the value in using the potentially richer api the "backend" offers, the design goal of metrics-api (and log-api fwiw) is that end users wont need to worry about the underlying "backend". using the "backend" api  is not even an option for libraries that integrate with these apis as they are not exposed to the "backend" which is dictated by the application, but the idea is also true for applications in vast majority of the cases. as such, i think we should not suggest switching to the "backend" api here as it send the opposite message. how about we add a section in the readme that explains this? /cc @ktoso @mrlotu

tomerd replied:
@ktoso @MrLotU wdyt?

ktoso replied:
Whoops seems I missed some emails last week here; Yes, agreed @tomerd 
The API here should not go around and make advertisements for what libraries can't use anyway 👍 

Amending now

ktoso commented:
Thanks for review everyone, sorry I missed it for a week; 
Addressed your feedback @tomerd, removed the sentence explaining what backends "might" do, agree that wrong place to pitch this: https://github.com/apple/swift-metrics/pull/35/commits/ee94fe5104b7d50001471cc2a6f47359500788d8
